core/src/main/scala/cromwell/core/simpleton/WdlValueBuilder.scala

```diff
@@ -0,0 +1,145 @@
+    // into the array.  e.g. for a component
+    // SimpletonComponent("[0][1]", v) this would return (0 -> SimpletonComponent("[1]", v)).
+    def descendIntoArray(component: SimpletonComponent): (Int, SimpletonComponent) = {
+      component.path match { case ArrayElementPattern(_, index, more) => index.toInt -> component.copy(path = more)}
+    }
+
+    // Returns a tuple of the key into the outermost map and a `SimpletonComponent` whose path reflects the "descent"
+    // into the map.  e.g. for a component
+    // SimpletonComponent(":bar:baz", v) this would return ("bar" -> SimpletonComponent(":baz", v)).
+    // Map keys are treated as Strings by this method, the caller must ultimately do the appropriate coercion to the
+    // actual map key type.
+    def descendIntoMap(component: SimpletonComponent): (String, SimpletonComponent) = {
+      component.path match { case MapElementPattern(_, key, more) => key.unescapeMeta -> component.copy(path = more)}
+    }
+
+    // Group tuples by key using a Map with key type `K`.
+    def group[K](tuples: Traversable[(K, SimpletonComponent)]): Map[K, Traversable[SimpletonComponent]] = {
+      tuples groupBy { case (i, _) => i } mapValues { _ map { case (i, s) => s} }
+    }
+
+    outputType match {
+      case _: WdlPrimitiveType => components collectFirst { case SimpletonComponent(_, v) => v } get
+      case arrayType: WdlArrayType =>
+        val groupedByArrayIndex: Map[Int, Traversable[SimpletonComponent]] = group(components map descendIntoArray)
+        WdlArray(arrayType, groupedByArrayIndex.toList.sortBy(_._1) map { case (_, s) => toWdlValue(arrayType.memberType, s) })
+      case mapType: WdlMapType =>
+        val groupedByMapKey: Map[String, Traversable[SimpletonComponent]] = group(components map descendIntoMap)
+        // map keys are guaranteed by the WDL spec to be primitives, so the "coerceRawValue(..).get" is safe.
+        WdlMap(mapType, groupedByMapKey map { case (k, ss) => mapType.keyType.coerceRawValue(k).get -> toWdlValue(mapType.valueType, ss) })
+    }
+  }
+
+  /** The two fields of the `SimpletonComponent` product type have types identical to those of the `WdlValueSimpleton`
+    * product type, but a `SimpletonComponent` is conceptually different from a `WdlValueSimpleton`.
+    * `SimpletonComponent` removes the outermost "name" piece of the `WdlValueSimpleton` key, so `path` contains only the
+    * path to the element.  e.g. for a `WdlValueSimpleton` of
+    *
+    * {{{
+    * WdlValueSimpleton("foo:bar[0]", WdlString("baz"))
+    * }}}
+    *
+    * the corresponding `SimpletonComponent` would be
+    *
+    * {{{
+    * SimpletonComponent(":bar[0]", WdlString("baz"))
+    * }}}
+    */
+  private case class SimpletonComponent(path: String, value: WdlValue)
+
+  def build(taskOutputs: Traversable[TaskOutput], simpletons: Traversable[WdlValueSimpleton]): Map[String, WdlValue] = {
```

Comment:
To me, it was the fact that this object exists for very specific use case of converting simpletons to JobOutputs and thus its should do that last bit, since it would be the remaining transformation. But if Simpleton -> WdlValue conversion could be reused elsewhere, then I see the use of leaving things generalized. I certainly don't feel strongly one way or another, just philosophizing.